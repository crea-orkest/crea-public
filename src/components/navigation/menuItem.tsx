import React from 'react'

export interface MenuItemProps {
  slug?: string | null
  label?: string | null
}

export const MenuItem: React.FC<MenuItemProps> = ({ slug, label }) => {
  if (!slug) return <li>No link specified</li>
  if (!label) return <li>No label specified</li>

  // should be a reusable function and simple to configure
  if (slug === 'homepage') {
    slug = '/'
  }

  return (
    <li>
      <a href={slug}>{label}</a>
    </li>
  )
}
