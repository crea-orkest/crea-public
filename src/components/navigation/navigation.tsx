import React from 'react'
import Link from 'next/link'
import classNames from 'classnames'
import type { GetGeneralInfoQuery } from 'graphql/generated/graphql'
import { NavigationItem } from 'components/navigationItem'
import { NavigationSubMenu } from 'components/navigationSubMenu'
import { MobileMenu } from 'components/mobileMenu'
import { CreaOrkestLogo } from 'components/icons/crea-orkest-logo'

import styles from './styles.module.scss'

interface Props {
  generalInfo: GetGeneralInfoQuery['general']
  siteName: string
}

export const Navigation = ({ siteName, generalInfo }: Props) => {
  return (
    <nav className={classNames(styles.root, 'content-layout')}>
      <div className={classNames(styles.content)}>
        <Link href="/" className={classNames(styles.logoLink)}>
          <CreaOrkestLogo className={classNames(styles.logo)} />
          <span className="sr-only">{siteName}: Ga naar home</span>
        </Link>
        <MobileMenu menu={generalInfo?.menu ?? null} />
        <ul className={classNames(styles.list)}>
          {generalInfo?.menu?.map((item) => {
            if ('link' in item) {
              return (
                <NavigationItem
                  key={item.id}
                  slug={item?.link?.slug}
                  label={item.label}
                />
              )
            }

            if ('menu' in item) {
              return (
                <NavigationSubMenu
                  key={item.id}
                  label={item.label}
                  submenu={item.menu}
                />
              )
            }

            return <NavigationItem key={item.id} slug={'/'} label={'Error'} />
          })}
        </ul>
      </div>
    </nav>
  )
}
