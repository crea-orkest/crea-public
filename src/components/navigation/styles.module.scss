.wrapper {
  max-width: 60rem;
  margin: 0 auto;
}

.list {
  display: flex;
  list-style: none;
  gap: 1rem;
  padding: 0;
}
