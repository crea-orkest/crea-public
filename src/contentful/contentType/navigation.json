{
  "sys": {
    "space": {
      "sys": { "type": "Link", "linkType": "Space", "id": "pwuxt9271wes" }
    },
    "id": "navigation",
    "type": "ContentType",
    "createdAt": "2023-07-01T09:14:40.236Z",
    "updatedAt": "2023-07-01T09:14:40.562Z",
    "environment": {
      "sys": { "id": "master", "type": "Link", "linkType": "Environment" }
    },
    "publishedVersion": 1,
    "publishedAt": "2023-07-01T09:14:40.562Z",
    "firstPublishedAt": "2023-07-01T09:14:40.562Z",
    "createdBy": {
      "sys": {
        "type": "Link",
        "linkType": "User",
        "id": "7dyBSKYa3yVW3xUX4g19qy"
      }
    },
    "updatedBy": {
      "sys": {
        "type": "Link",
        "linkType": "User",
        "id": "7dyBSKYa3yVW3xUX4g19qy"
      }
    },
    "publishedCounter": 1,
    "version": 2,
    "publishedBy": {
      "sys": {
        "type": "Link",
        "linkType": "User",
        "id": "7dyBSKYa3yVW3xUX4g19qy"
      }
    }
  },
  "displayField": "title",
  "name": "navigation",
  "description": "",
  "fields": [
    {
      "id": "title",
      "name": "title",
      "type": "Symbol",
      "localized": true,
      "required": true,
      "validations": [],
      "disabled": false,
      "omitted": false
    },
    {
      "id": "logo",
      "name": "logo",
      "type": "Link",
      "localized": false,
      "required": true,
      "validations": [{ "linkMimetypeGroup": ["image"] }],
      "disabled": false,
      "omitted": false,
      "linkType": "Asset"
    },
    {
      "id": "navigationItems",
      "name": "navigation items",
      "type": "Array",
      "localized": false,
      "required": false,
      "validations": [
        { "size": { "min": 0, "max": 3 }, "message": "max 3 entries" }
      ],
      "disabled": false,
      "omitted": false,
      "items": {
        "type": "Link",
        "validations": [{ "linkContentType": ["page"] }],
        "linkType": "Entry"
      }
    }
  ]
}
