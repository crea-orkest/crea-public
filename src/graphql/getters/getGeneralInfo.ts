import { client } from '../gqlClient'
import {
  GetGeneralInfoDocument,
  type GetGeneralInfoQuery,
  type GetGeneralInfoQueryVariables,
} from '../generated/graphql'

export const getGeneralInfo = async () => {
  try {
    const { data } = await client.query<
      GetGeneralInfoQuery,
      GetGeneralInfoQueryVariables
    >(GetGeneralInfoDocument, {})

    console.log('General info fetched', data)

    return {
      data,
    }
  } catch (error) {
    if (error instanceof Error) console.log(error.message)
    return { data: null }
  }
}
