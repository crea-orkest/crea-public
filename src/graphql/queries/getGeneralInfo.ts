import { generalInfo } from 'graphql/fragments/generalInfo'
import { gql } from '@urql/core'

export const GET_GENERAL_INFO = gql`
  query getGeneralInfo {
    general {
      ...generalInfo
    }
  }
  ${generalInfo}
`
