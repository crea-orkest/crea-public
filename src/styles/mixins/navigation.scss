@use 'sass:map';
@use '../tokens/colors.scss' as colors;
@use '../tokens/typography.scss' as typography;
@use '../tokens/spacing.scss' as spacing;

@mixin link {
  display: block;
  color: colors.$dark;
  font-family: typography.$dm-serif;
  padding-top: map.get(spacing.$rem, 8);
  padding-bottom: map.get(spacing.$rem, 8);

  &:hover,
  &:focus {
    text-decoration: none;
    color: colors.$secondary;
  }
}
