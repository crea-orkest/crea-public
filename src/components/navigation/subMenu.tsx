'use client'

import { MenuItem } from './menuItem'
import React from 'react'
import type { SubmenuItemFragment } from 'graphql/generated/graphql'

export interface SubMenuProps {
  label?: string | null
  item: string
}

export const SubMenu: React.FC<SubMenuProps> = ({ label, item }) => {
  const submenu: SubmenuItemFragment['menu'] = JSON.parse(item)
  const [visible, setVisible] = React.useState(false)

  const handleClick = () => {
    setVisible(!visible)
  }

  if (!label) return null

  return (
    <li>
      <button
        disabled={Boolean(submenu?.length === 0)}
        type="button"
        onClick={handleClick}
      >
        {label}
      </button>
      {visible && (
        <ul>
          {submenu?.map((item) => (
            <MenuItem
              key={item.id}
              slug={item?.link?.slug}
              label={item.label}
            />
          ))}
        </ul>
      )}
    </li>
  )
}
