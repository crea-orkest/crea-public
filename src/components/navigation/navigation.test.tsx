import { Navigation } from './navigation'
import React from 'react'
import { getGeneralInfo } from 'graphql/getters/getGeneralInfo'
import { mockMenuData } from './mockData'
import { resolvedComponent } from '../../testHelpers/resolvedComponent'
import { render, screen } from '@testing-library/react'

jest.mock('../../graphql/getters/getGeneralInfo', () => {
  return {
    __esModule: true,
    getGeneralInfo: jest.fn(),
  }
})

const getGeneralInfoMock = jest.mocked(getGeneralInfo)

describe('Concert component', () => {
  it('shows all the data', async () => {
    getGeneralInfoMock.mockResolvedValueOnce({
      data: mockMenuData,
      error: undefined,
    })

    const Resolved = await resolvedComponent(Navigation, {})

    render(<Resolved />)

    expect(screen.getByText('contact')).toBeTruthy()
  })
})
