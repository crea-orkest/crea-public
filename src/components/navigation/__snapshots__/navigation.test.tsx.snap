// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Concert component shows all the data 1`] = `
<div>
  <nav
    class="wrapper"
  >
    <ul
      class="list"
    >
      <li
        class="navItem"
      >
        <a
          href="/"
        >
          Homepage
        </a>
      </li>
      <li
        class="navItem"
      >
        <a
          href="/concerten"
        >
          Concerten
        </a>
      </li>
      <li
        class="navItem"
      >
        <a
          href="/contact"
        >
          contact
        </a>
      </li>
      <li
        class="navItem"
      >
        <a
          href="/"
        >
          homepage
        </a>
      </li>
      <li>
        <button
          type="button"
        >
          level 1
        </button>
      </li>
    </ul>
  </nav>
</div>
`;
